### PR TITLE
Harmony: Autocomplete

### LINKED ISSUE #669: Harmony: Autocomplete
image
image

### PR DESCRIPTION
**Pull request by LamaEats**
... and related components and fixes

## PR includes

- [x] Bug Fix
- [x] Feature

## Related issues

Resolve #669 

## QA Instructions, Screenshots, Recordings
  
In this pr includes the same implementation of early create components

### CODE REVIEW

File: src/harmony/AutoComplete/AutoComplete.tsx [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+                className={classNames(classes.AutoCompleteRadioGroup, className)}
+                name={name}
+                onChange={handleChange}
+                value={value}
+                {...attrs}
+            >
+                {items.map((item) => (
+                    <Radio className={classes.AutoCompleteRadio} value={item.value}>
+                        <RadioLabel>
+                            <span>{item.title}</span>
```
Comment (awinogradov) [inline src/harmony/AutoComplete/AutoComplete.tsx:119]: Do we really need span here?

File: src/harmony/RadioGroup/RadioGroup.tsx [hunk 2]
```diff
@@ -0,0 +1,108 @@
+import React, { HTMLAttributes, PropsWithChildren, createContext, useCallback, useId } from 'react';
```
Comment (awinogradov) [inline src/harmony/RadioGroup/RadioGroup.tsx:1]: useCallback is never used

File: src/harmony/RadioGroup/RadioGroup.tsx [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+    className,
+    children,
+    ...attrs
+}) => {
+    const id = useId();
+
+    return (
+        <RadioGroupContext.Consumer>
+            {({ name, onChange, value: selected }) => (
+                <RadioContext.Provider value={{ id: `${id}-${name}`, value, onChange, name, disabled }}>
```
Comment (awinogradov) [inline src/harmony/RadioGroup/RadioGroup.tsx:77]: memo

**Review by awinogradov — CHANGES_REQUESTED** [review]

File: src/harmony/AutoComplete/AutoComplete.tsx (hunk 1 above)
Comment (LamaEats) [reply src/harmony/AutoComplete/AutoComplete.tsx:119]: Perhaps, don't needed

File: src/harmony/AutoComplete/AutoComplete.tsx [hunk 4]
```diff
@@ -0,0 +1,277 @@
+import React, { createContext, useContext, useCallback, useMemo, useEffect, useState, useRef, memo } from 'react';
+import classNames from 'classnames';
+
+import { nullable } from '../../utils';
+import { Input } from '../Input/Input';
+import { ListViewItem, ListView } from '../../components/ListView/ListView';
+import { RadioControl as Radio, RadioGroup, RadioGroupLabel, RadioLabel } from '../RadioGroup/RadioGroup';
```
Comment (awinogradov) [inline src/harmony/AutoComplete/AutoComplete.tsx:7]: 'RadioLabel' is defined but never used

File: src/harmony/AutoComplete/AutoComplete.tsx [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+type InputProps = React.ComponentProps<typeof Input>;
+type AutoCompleteMode = 'single' | 'multiple';
+type AutoCompleteSelectedMap = Set<string>;
+type ListItemProps = Parameters<React.ComponentProps<typeof ListViewItem>['renderItem']>[0];
+
+interface AutoCompleteRenderItemProps<T> {
+    item: T;
+    index: number;
+    onItemClick: () => void;
+    checked: boolean;
```
Comment (awinogradov) [inline src/harmony/AutoComplete/AutoComplete.tsx:20]: optional?

File: src/harmony/Checkbox/Checkbox.tsx [hunk 6]
```diff
@@ -11,35 +11,29 @@ interface CheckboxProps extends React.HTMLAttributes<HTMLInputElement> {
...
+            className={cn(classes.Checkbox, className, {
+                [classes.CheckboxDisabled]: disabled,
+            })}
+        >
+            <input
+                className={cn(classes.CheckboxInput)}
+                type="checkbox"
+                name={name}
+                defaultChecked={checked}
+                onChange={onChange}
```
Comment (awinogradov) [inline src/harmony/Checkbox/Checkbox.tsx:29]: onChange, name, disabled are exist in rest

File: src/harmony/Radio/Radio.tsx [hunk 7]
```diff
@@ -11,35 +11,29 @@ interface RadioProps extends React.HTMLAttributes<HTMLInputElement> {
...
+            className={cn(classes.Radio, className, {
+                [classes.RadioDisabled]: disabled,
+            })}
+        >
+            <input
+                className={cn(classes.RadioInput)}
+                type="radio"
+                name={name}
+                defaultChecked={checked}
+                onChange={onChange}
```
Comment (awinogradov) [inline src/harmony/Radio/Radio.tsx:29]: Same comment with Checkbox

File: src/harmony/RadioGroup/RadioGroup.tsx [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+    children,
+    ...attrs
+}) => {
+    const id = useId();
+    const { name, onChange, value: selected } = useContext(RadioGroupContext);
+
+    return (
+        <Radio
+            className={cn(classes.RadioControl, className)}
+            name={name}
```
Comment (awinogradov) [inline src/harmony/RadioGroup/RadioGroup.tsx:52]: Same comment with Checkbox and Radio

**Review by awinogradov — CHANGES_REQUESTED** [review]

File: src/harmony/AutoComplete/AutoComplete.tsx [hunk 9]
```diff
@@ -0,0 +1,277 @@
...
+}
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+const AutoCompleteContextProvider = createContext<AutoCompleteContext<any> | null>(null);
+
+function useAutoCompleteContext<T extends { id: string }>(): AutoCompleteContext<T> {
+    const ctx = useContext(AutoCompleteContextProvider) as AutoCompleteContext<T> | null;
+
+    if (!ctx) {
+        throw new Error("Don't use before initialization or outse of `AutoComplete` component");
```
Comment (awinogradov) [inline src/harmony/AutoComplete/AutoComplete.tsx:78]: outse -> outside

**Review by awinogradov — APPROVED** [review]

**Review by awinogradov — CHANGES_REQUESTED** [review]
RadioGroup not fixed

File: src/harmony/AutoComplete/AutoComplete.stories.tsx [hunk 10]
```diff
@@ -0,0 +1,256 @@
...
+//                 value={value}
+//                 onChange={setValue}
+//                 iconLeft={<IconSearchOutline size="s" />}
+//                 placeholder="Type `tit...` for search"
+//             />
+//             <AutoCompleteList selected />
+//             <AutoCompleteList filterSelected title="Sugesstions" />
+//         </AutoComplete>
+//     );
+// };
```
[the remark this answers is not in the document]
Comment (LamaEats) [reply src/harmony/AutoComplete/AutoComplete.stories.tsx:224]: for this https://taskany-inc.github.io/bricks/?path=/story/autocomplete--multiple-with-table
but harmony lib haven't Table components now

**Review by 9teen90nine — APPROVED** [review]

**Pull request merged**